src/main/java/com/shapesecurity/salvation/Parser.java

```diff
@@ -10,82 +9,141 @@
+        return new Parser(Tokeniser.tokenise(sourceText), URI.parse(origin), null).parsePolicyListAndAssertEOF();
     }
 
-    @Nonnull
-    public static List<Policy> parseMulti(@Nonnull String sourceText, @Nonnull Origin origin,
-        @Nonnull Collection<Warning> warningsOut) throws ParseException, TokeniserException {
-        return new Parser(Tokeniser.tokenise(sourceText), origin, warningsOut)
-            .parsePolicyListAndAssertEOF();
+    @Nonnull public static List<Policy> parseMulti(@Nonnull String sourceText, @Nonnull Origin origin,
+        @Nonnull Collection<Notice> warningsOut) {
+        return new Parser(Tokeniser.tokenise(sourceText), origin, warningsOut).parsePolicyListAndAssertEOF();
     }
 
-    @Nonnull
-    public static List<Policy> parseMulti(@Nonnull String sourceText, @Nonnull String origin,
-        @Nonnull Collection<Warning> warningsOut) throws ParseException, TokeniserException {
-        return new Parser(Tokeniser.tokenise(sourceText), URI.parse(origin), warningsOut)
-            .parsePolicyListAndAssertEOF();
+    @Nonnull public static List<Policy> parseMulti(@Nonnull String sourceText, @Nonnull String origin,
+        @Nonnull Collection<Notice> warningsOut) {
+        return new Parser(Tokeniser.tokenise(sourceText), URI.parse(origin), warningsOut).parsePolicyListAndAssertEOF();
+    }
+
+    @Nonnull private static String trimRHSWS(@Nonnull String s) {
+        int i;
+        for (i = s.length() - 1; i >= 0; --i) {
+            int c = s.codePointAt(i);
+            if (WSP.matcher(new String(new int[]{c}, 0, 1)).find())
+                return s.substring(0, i + 1);
+        }
+        return s;
     }
 
-    @Nonnull protected Warning createWarning(@Nonnull String message) {
-        return new Warning(message);
+    @Nonnull protected Notice createNotice(@Nonnull Notice.Type type, @Nonnull String message) {
+        return new Notice(type, message);
+    }
+
+    @Nonnull protected Notice createNotice(@Nullable Token token, @Nonnull Notice.Type type, @Nonnull String message) {
+        return new Notice(type, message);
     }
 
     private void warn(@Nonnull String message) {
-        if (this.warningsOut != null) {
-            this.warningsOut.add(this.createWarning(message));
+        this.warn(this.getCurrentToken(), message);
+    }
+
+    private void warn(@Nullable Token token, @Nonnull String message) {
+        if (this.noticesOut != null) {
+            this.noticesOut.add(this.createNotice(token, Notice.Type.WARNING, message));
+        }
+    }
+
+    private void error(@Nonnull String message) {
+        this.error(this.getCurrentToken(), message);
+    }
+
+    private void error(@Nullable Token token, @Nonnull String message) {
+        if (this.noticesOut != null) {
+            this.noticesOut.add(this.createNotice(token, Notice.Type.ERROR, message));
+        }
+    }
+
+    private void info(@Nonnull String message) {
```

Comment:
Are these notice functions that work on the current token used at all? If we can, we should try to drop them.